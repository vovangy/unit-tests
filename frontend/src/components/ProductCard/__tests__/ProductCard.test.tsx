import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { ProductCard } from '../ProductCard';
import { getPrice } from '../../../utils';

afterEach(jest.clearAllMocks);
jest.mock('../../../utils');

describe('ProductCard component tests', () => {
    it('should mock getPrice function and display correct formatted price', () => {
        jest.mocked(getPrice).mockReturnValue('200 $');

        const rendered = render(
            <ProductCard
                id={101}
                name={'Samsung Galaxy S21'}
                description={'Latest Samsung smartphone'}
                price={200}
                category={'Электроника'}
                priceSymbol={'$'}
            />
        );

        expect(rendered.asFragment()).toMatchSnapshot();

        expect(getPrice).toHaveBeenCalledTimes(1);
        expect(getPrice).toHaveBeenCalledWith(200, '$');
        expect(getPrice).lastReturnedWith('200 $');
    });

    it('should render the product image with given URL', () => {
        const rendered = render(
            <ProductCard
                id={102}
                name={'Dell XPS 13'}
                description={'Ultra-slim laptop'}
                price={1500}
                category={'Электроника'}
                imgUrl={'public/dell-xps.png'}
            />
        );

        expect(rendered.asFragment()).toMatchSnapshot();
        expect(rendered.getByAltText('Dell XPS 13')).toHaveClass(
            'product-card__image'
        );
    });

    it('should not render image if imgUrl is not provided', () => {
        const rendered = render(
            <ProductCard
                id={103}
                name={'Sony WH-1000XM4'}
                description={'Noise-cancelling headphones'}
                price={300}
                priceSymbol={'$'}
                category={'Электроника'}
            />
        );

        expect(rendered.asFragment()).toMatchSnapshot();
        expect(rendered.queryAllByAltText('Sony WH-1000XM4')).toHaveLength(0);
        expect(rendered.queryByAltText('Sony WH-1000XM4')).not.toBeInTheDocument();
    });

});
