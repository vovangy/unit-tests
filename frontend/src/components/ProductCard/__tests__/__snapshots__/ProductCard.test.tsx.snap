// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard component tests should mock getPrice function and display correct formatted price 1`] = `
<DocumentFragment>
  <div
    class="product-card"
  >
    <div
      class="product-card__text"
    >
      <h2
        class="product-card__name"
      >
        Samsung Galaxy S21
      </h2>
      <div
        class="product-card__description"
      >
        Latest Samsung smartphone
      </div>
      <h1
        class="product-card__price"
      >
        200 $
      </h1>
      <div
        class="product-card__category"
      >
        Электроника
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProductCard component tests should not render image if imgUrl is not provided 1`] = `
<DocumentFragment>
  <div
    class="product-card"
  >
    <div
      class="product-card__text"
    >
      <h2
        class="product-card__name"
      >
        Sony WH-1000XM4
      </h2>
      <div
        class="product-card__description"
      >
        Noise-cancelling headphones
      </div>
      <h1
        class="product-card__price"
      >
        200 $
      </h1>
      <div
        class="product-card__category"
      >
        Электроника
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProductCard component tests should render the product image with given URL 1`] = `
<DocumentFragment>
  <div
    class="product-card"
  >
    <div
      class="product-card__text"
    >
      <h2
        class="product-card__name"
      >
        Dell XPS 13
      </h2>
      <div
        class="product-card__description"
      >
        Ultra-slim laptop
      </div>
      <h1
        class="product-card__price"
      >
        200 $
      </h1>
      <div
        class="product-card__category"
      >
        Электроника
      </div>
    </div>
    <img
      alt="Dell XPS 13"
      class="product-card__image"
      src="public/dell-xps.png"
    />
  </div>
</DocumentFragment>
`;
